File: payView/DWPayView/DWEnterPasswordView.m
//
//  DWEnterPasswordView.m
//  payView
//
//  Created by DA WENG on 2017/1/9.
//  Copyright © 2017年 DA WENG. All rights reserved.
//

#import "DWEnterPasswordView.h"
#import "DWSizeDefined.h"
#import "DWNumberKeyboardView.h"
#import "DWManageData.h"
#import "DWStateView.h"
#define kDotSize CGSizeMake (10, 10) //密码点的大小
#define kDotCount 6  //密码个数
#define K_Field_Height 45  //每一个输入框的高度
@interface DWEnterPasswordView ()<UITextFieldDelegate,SafeKeyBoardDelegate,ManageDataDelegate>
{
    DWStateView* stateView;
    UILabel* stateLabel;
    
}

@property (nonatomic,strong) NSMutableArray *dotArray; //用于存放黑色的点点
@property(nonatomic,strong)DWNumberKeyboardView* numberKeyboard;
@property(nonatomic,strong)UIView* passWordView;

@end

@implementation DWEnterPasswordView


/*
 代理传值获取到的交易结果
 只判断success或者fail来显示动画。
 */
-(void)passData:(id)data{
    NSLog(@"DWEnterPasswordView:%@",data);
    NSDictionary* dict = [[NSDictionary alloc]init];
    dict = data;
    stateView.hidden = YES;
    NSString* result = [dict objectForKey:@"result"];
    if ([result isEqualToString:@"success"]) {
        DWStateView* tickState = [[DWStateView alloc]initWithFrame:CGRectMake(self.frame.size.width*3/8, self.frame.size.width/4,self.frame.size.width/4,self.frame.size.width/4) withType:DWStateDisplayTypeSuccessTick withColor:[UIColor colorWithRed:53.0/255.0 green:203.0/255.0 blue:75.0/255.0 alpha:1]];
        [self addSubview:tickState];
        stateLabel.text = @"支付成功";
        [self performSelector:@selector(successDelayMethod) withObject:nil afterDelay:2.8f];
    }else if ([result isEqualToString:@"fail"]){
        
        DWStateView* tickState = [[DWStateView alloc]initWithFrame:CGRectMake(self.frame.size.width*3/8, self.frame.size.width/4,self.frame.size.width/4,self.frame.size.width/4) withType:DWStateDisplayTypeFailCross withColor:[UIColor colorWithRed:252.0/255.0 green:99.0/255.0 blue:94.0/255.0 alpha:1]];
        [self addSubview:tickState];
        stateLabel.text = @"支付失败";
        [self performSelector:@selector(failDelayMethod) withObject:nil afterDelay:2.8f];
        
    }
    
}

//动画显示完成后，等0.8秒整个页面消失
-(void)successDelayMethod{
//    stateLabel.text = @"支付成功";
    [self performSelector:@selector(postNotification) withObject:nil afterDelay:0.8f];
    
}

//动画显示完成后，等0.8秒整个页面消失
-(void)failDelayMethod{
//    stateLabel.text = @"支付失败";
    [self performSelector:@selector(postNotification) withObject:nil afterDelay:0.8f];
}

//通知DWPayView，回收页面
-(void)postNotification{
    [[NSNotificationCenter defaultCenter] postNotificationName:@"finishPay" object:nil userInfo:nil];
}

#pragma mark - init
- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        
        self.backgroundColor = [UIColor whiteColor];
        
        self.numberKeyboard = [[DWNumberKeyboardView alloc]init];
        self.numberKeyboard.delegate = self;
        self.textField.inputView = self.numberKeyboard;
        [self addSubview:self.textField];
        [self.textField becomeFirstResponder];
        [self initPwdTextField];
        
        
    }
    return self;
}

- (void)initPwdTextField
{
    //每个密码输入框的宽度
    CGFloat width = (dwDEVICESCREENWIDTH - dwMARGINTOLEFT*2) / kDotCount;
    self.passWordView = [[UIView alloc]initWithFrame:self.textField.frame];
    [self addSubview:self.passWordView];
    //生成分割线
    for (int i = 0; i < kDotCount - 1; i++) {
        UIView *lineView = [[UIView alloc] initWithFrame:CGRectMake(0 + (i + 1) * width, 0, 1, K_Field_Height)];
        lineView.backgroundColor = [UIColor grayColor];
        [self.passWordView addSubview:lineView];
    }
    
    self.dotArray = [[NSMutableArray alloc] init];
    //生成中间的点
    for (int i = 0; i < kDotCount; i++) {
        UIView *dotView = [[UIView alloc] initWithFrame:CGRectMake(0 + (width - kDotCount) / 2 + i * width, 0 + (K_Field_Height - kDotSize.height) / 2, kDotSize.width, kDotSize.height)];
        dotView.backgroundColor = [UIColor blackColor];
        dotView.layer.cornerRadius = kDotSize.width / 2.0f;
        dotView.clipsToBounds = YES;
        dotView.hidden = YES; //先隐藏
        [self.passWordView addSubview:dotView];
        //把创建的黑色点加入到数组中
        [self.dotArray addObject:dotView];
    }
    
}

- (UITextField *)textField
{
    if (!_textField) {
        _textField = [[UITextField alloc] initWithFrame:CGRectMake(dwMARGINTOLEFT, dwMARGINTOLEFT, dwDEVICESCREENWIDTH - dwMARGINTOLEFT*2, K_Field_Height)];
        _textField.backgroundColor = [UIColor whiteColor];
        //输入的文字颜色为白色
        _textField.textColor = [UIColor whiteColor];
        //输入框光标的颜色为白色
        _textField.tintColor = [UIColor whiteColor];
        _textField.delegate = self;
        _textField.autocapitalizationType = UITextAutocapitalizationTypeNone;
        _textField.keyboardType = UIKeyboardTypeNumberPad;
        _textField.layer.borderColor = [[UIColor grayColor] CGColor];
        _textField.layer.borderWidth = 1;
        [_textField addTarget:self action:@selector(inputtext:) forControlEvents:UIControlEventEditingChanged];//使用自定义键盘
        //         [_textField addTarget:self action:@selector(textFieldDidChange:) forControlEvents:UIControlEventEditingChanged];使用系统键盘
    }
    return _textField;
}

-(void)inputtext:(NSString *)inputtext{
    self.textField.text = inputtext;
    for (UIView *dotView in self.dotArray) {
        dotView.hidden = YES;
    }
    for (int i = 0; i < self.textField.text.length; i++) {
        
        ((UIView *)[self.dotArray objectAtIndex:i]).hidden = NO;
    }
    if (self.textField.text.length == kDotCount) {
        
        [[NSNotificationCenter defaultCenter] postNotificationName:@"disableBackBtn" object:nil userInfo:nil];//禁用返回键通知
        
        [self.numberKeyboard removeFromSuperview];
        [self.textField removeFromSuperview];
        [self.passWordView removeFromSuperview];
        
        stateView = [[DWStateView alloc]initWithFrame:CGRectMake(self.frame.size.width*3/8, self.frame.size.width/4,self.frame.size.width/4,self.frame.size.width/4) withType:DWStateDisplayTypeDrawCircle withColor:[UIColor colorWithRed:53.0/255.0 green:203.0/255.0 blue:75.0/255.0 alpha:1]];
        [self addSubview:stateView];
        stateLabel = [[UILabel alloc]initWithFrame:CGRectMake(stateView.frame.origin.x, stateView.frame.origin.y+stateView.frame.size.height+10, stateView.frame.size.width, stateView.frame.size.height/4)];
        stateLabel.text = @"正在支付";
        stateLabel.textAlignment = NSTextAlignmentCenter;
        [self addSubview:stateLabel];
        
        
        DWManageData* manageData = [[DWManageData alloc]init];
        manageData.delegate = self;
        [manageData payOrderWithPassword:inputtext withBankCardInfo:self.payBankInfo];
    }
    
}


//使用系统自带键盘
- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string
{
    NSLog(@"变化%@", string);
    if([string isEqualToString:@"\n"]) {
        //按回车关闭键盘
        [textField resignFirstResponder];
        return NO;
    } else if(string.length == 0) {
        //判断是不是删除键
        return YES;
    }
    else if(textField.text.length >= kDotCount) {
        //输入的字符个数大于6，则无法继续输入，返回NO表示禁止输入
        NSLog(@"输入的字符个数大于6，忽略输入");
        return NO;
    } else {
        return YES;
    }
}


- (void)textFieldDidChange:(UITextField *)textField
{
    NSLog(@"%@", textField.text);
    for (UIView *dotView in self.dotArray) {
        dotView.hidden = YES;
    }
    for (int i = 0; i < textField.text.length; i++) {
        ((UIView *)[self.dotArray objectAtIndex:i]).hidden = NO;
    }
    if (textField.text.length == kDotCount) {
        NSLog(@"输入完毕");
    }
}


@end

File: DWPayViewSwift/DWPayViewSwift/DWEnterPasswordView.swift
//
//  DWEnterPasswordView.swift
//  DWPayViewSwift
//
//  Created by DA WENG on 2017/2/24.
//  Copyright © 2017年 DA WENG. All rights reserved.
//

import UIKit

let DOTSIZE = CGSize(width:10, height:10) //密码点的大小
let DOTCOUNT = 6  //密码个数
let FIELDHEIGHT = 45  //每一个输入框的高度

class DWEnterPasswordView: UIView ,DWNumberKeyBoardDelegate,DWManageDataDelegate{
    var loadingView = DWSateView()
    var stateLabel = UILabel()
    var dotArray = NSMutableArray()
    let numberKeyboard = DWNumberKeyboardView()
    let passWordView = UIView()
    var payBankInfo = NSDictionary()
    
    override init(frame:CGRect){
        
        super.init(frame: frame)
        
        self.backgroundColor = UIColor.white
        self.numberKeyboard.delegate = self
        textField.inputView = self.numberKeyboard
        textField .becomeFirstResponder()
        self.addSubview(textField)
        self.initPwdTextField()
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func passData(data: Any) {
        
        var dict = NSDictionary()
        dict = data as! NSDictionary
        loadingView.removeFromSuperview()
        let result = dict.object(forKey: "result") as! String
        if result == "success" {
            let tickState = DWSateView(frame: CGRect(x: self.frame.size.width*3/8, y: self.frame.size.width/4, width: self.frame.size.width/4, height: self.frame.size.width/4), color: UIColor.init(red: 53.0/255.0, green: 203.0/255.0, blue: 75.0/255.0, alpha: 1), type: DWStateDisplayType.successTickWithFullCoolor)
            self.addSubview(tickState)
            stateLabel.text = "支付成功"
            self.perform(#selector(successDelayMethod), with: nil, afterDelay: 2.8)
        }else if result == "fail"{
        
            let crossState = DWSateView(frame: CGRect(x: self.frame.size.width*3/8, y: self.frame.size.width/4, width: self.frame.size.width/4, height: self.frame.size.width/4), color: UIColor.init(red: 252.0/255.0, green: 99.0/255.0, blue: 94.0/255.0, alpha: 1), type: DWStateDisplayType.failCrossWithFullCoolor)
            self.addSubview(crossState)
            stateLabel.text = "支付失败"
            self.perform(#selector(failDelayMethod), with: nil, afterDelay: 2.8)
            
        }

    }
    
    func successDelayMethod(){
    self.perform(#selector(postNotification), with: nil, afterDelay: 1.5)
    
    }
    
    func failDelayMethod(){
    
        self.perform(#selector(postNotification), with: nil, afterDelay: 1.5)
    }
    
    func postNotification(){

        NotificationCenter.default.post(name: NSNotification.Name(rawValue: "finishPay"), object: nil)
    }
    
    let textField: UITextField = {
        let textField = UITextField(frame: CGRect(x: dwMARGINTOLEFT, y: dwMARGINTOLEFT, width: dwDEVICESCREENWIDTH - dwMARGINTOLEFT*2, height: CGFloat(FIELDHEIGHT)))
        textField.backgroundColor = UIColor.white
        textField.textColor = UIColor.white
        textField.tintColor = UIColor.white
        textField.autocapitalizationType = UITextAutocapitalizationType.none
        textField.layer.borderColor = UIColor.gray.cgColor
        textField.layer.borderWidth = 1
        textField.addTarget(self, action: #selector(inputText), for: UIControlEvents.editingChanged)
        return textField
    }()
    
    
    func inputText(inputText: String) {
        
        
        for dotView in self.dotArray {
            (dotView as! UIView).isHidden = true
        }
        for i in 0..<inputText.characters.count {
            (self.dotArray.object(at: i) as! UIView).isHidden = false
        }
        if inputText.characters.count == DOTCOUNT {
            NotificationCenter.default.post(name: NSNotification.Name(rawValue: "disableBackBtn"), object: nil)
            self.numberKeyboard.removeFromSuperview()
            self.textField.removeFromSuperview()
            self.passWordView.removeFromSuperview()
            
            loadingView = DWSateView(frame: CGRect(x: self.frame.size.width*3/8, y: self.frame.size.width/4, width: self.frame.size.width/4, height: self.frame.size.width/4), color: UIColor.init(red: 53.0/255.0, green: 203.0/255.0, blue: 75.0/255.0, alpha: 1), type: DWStateDisplayType.drawCircle)
            self.addSubview(loadingView)
            
            self.stateLabel = UILabel(frame: CGRect(x: self.frame.size.width*3/8, y: self.frame.size.width/4+self.frame.size.width/4+10, width: self.frame.size.width/4, height: self.frame.size.width/16))
            stateLabel.text = "正在支付"
            stateLabel.textAlignment = NSTextAlignment.center
            self.addSubview(stateLabel)
        
            let manageData = DWManageData()
            manageData.delegate = self
            manageData.payOrder(withPassword: inputText, bankCardInfo: self.payBankInfo)
   
        }
    }
    
    func initPwdTextField(){
        
        let width:CGFloat = (dwDEVICESCREENWIDTH - dwMARGINTOLEFT*2)/CGFloat(DOTCOUNT)
        self.passWordView.frame = self.textField.frame
        self.addSubview(self.passWordView)
        
        for i in 0..<DOTCOUNT-1 {
            let lineView = UIView(frame: CGRect(x: (i + 1) * Int(width), y: 0, width: 1, height: FIELDHEIGHT))
            lineView.backgroundColor = UIColor.gray
            self.passWordView.addSubview(lineView)
        }
        for i in 0...DOTCOUNT {
            let dotView = UIView(frame: CGRect(x: (width - CGFloat(DOTCOUNT)) / 2 + CGFloat(i) * width, y: (CGFloat(FIELDHEIGHT) - DOTSIZE.height) / 2, width: DOTSIZE.width, height:  DOTSIZE.height))
            dotView.backgroundColor = UIColor.black
            dotView.layer.cornerRadius = DOTSIZE.width/2
            dotView.clipsToBounds = true
            dotView.isHidden = true
            self.passWordView.addSubview(dotView)
            self.dotArray.add(dotView)
        }
        
    }

    
    
}
